src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -2759,6 +2766,17 @@ private BoundExpression BindArgumentExpression(DiagnosticBag diagnostics, Expres
                     Debug.Assert(parameterTypeWithAnnotations.HasType);
                     arguments[arg] = ((BoundDiscardExpression)argument).SetInferredTypeWithAnnotations(parameterTypeWithAnnotations);
                 }
+                else if (argument.NeedsToBeConverted())
+                {
+                    Debug.Assert(kind.IsIdentity);
+                    if (argument is BoundTupleLiteral sourceTuple)
+                    {
+                        TypeWithAnnotations parameterTypeWithAnnotations = GetCorrespondingParameterTypeWithAnnotations(ref result, parameters, arg);
+                        TupleTypeSymbol.ReportNamesMismatchesIfAny(parameterTypeWithAnnotations.Type, sourceTuple, diagnostics);
```

Comment:
>ReportNamesMismatchesIfAny [](start = 40, length = 26)

It seems strange to have a case for `!kind.IsIdentity` then a case for `argument.NeedsToBeConverted()`.
Could we run both cases through `CreateConversion` instead, to factor the code and not add a special case to call `ReportNamesMismatchesIfAny` here?

Reply:
There is an open issue around that.  Perhaps CreateConversion should create an identity conversion node when the types differ for example in tuple element names, but that would be a change of behavior so I don't want to do that here.

---
In reply to: [301827169](https://github.com/dotnet/roslyn/pull/37052#discussion_r301827169) [](ancestors = 301827169)